src/Http/Routing/src/Matching/DfaMatcherBuilder.cs

```diff
@@ -186,131 +189,257 @@ public DfaNode BuildDfaTree(bool includeLabel = false)
                             {
-                                for (var k = 0; k < parameterPolicyReferences.Count; k++)
+                                // We have a pattern like {a}-{b}.{part?} and a literal "a-b". Since we've matched the complex segment it means that the optional
+                                // parameter was not specified, so we skip it.
+                                Debug.Assert(i == segment.Parts.Count - 1 && partParameter.IsOptional);
+                                continue;
+                            }
+
+                            if (endpoint.RoutePattern.ParameterPolicies.TryGetValue(partParameter.Name, out var parameterPolicyReferences))
+                            {
+                                for (var j = 0; j < parameterPolicyReferences.Count; j++)
                                 {
-                                    var reference = parameterPolicyReferences[k];
+                                    var reference = parameterPolicyReferences[j];
                                     var parameterPolicy = _parameterPolicyFactory.Create(parameterPart, reference);
-                                    if (parameterPolicy is IOutboundParameterTransformer parameterTransformer)
+                                    if (parameterPolicy is ILiteralConstraint constraint && !constraint.MatchLiteral(partParameter.Name, (string)parameterValue))
                                     {
-                                        requiredValue = parameterTransformer.TransformOutbound(requiredValue);
+                                        passedAllPolicies = false;
                                         break;
                                     }
                                 }
                             }
+                        }
 
-                            var literalValue = requiredValue?.ToString() ?? throw new InvalidOperationException($"Required value for literal '{parameterPart.Name}' must evaluate to a non-null string.");
-
-                            AddLiteralNode(includeLabel, nextParents, parent, literalValue);
+                        if (passedAllPolicies)
+                        {
+                            nextParents.Add(parent.Literals[literal]);
                         }
-                        else if (segment.IsSimple && parameterPart != null)
+                    }
+
+                    routeValues.Clear();
+                }
+            }
+
+            private readonly void AddParentsWithMatchingLiteralConstraints(List<DfaNode> nextParents, DfaNode parent, RoutePatternParameterPart parameterPart, IReadOnlyList<RoutePatternParameterPolicyReference> parameterPolicyReferences)
+            {
+                // The list of parameters that fail to meet at least one ILiteralConstraint.
+                var hasFailingPolicy = parent.Literals.Keys.Count < 256 ?
+                    stackalloc bool [parent.Literals.Keys.Count] :
```

Comment:
Use a constant size plus slice with stackalloc. It is faster to use a constant size because the compiler knows the exact stack size required and can make optimizations.

Reply:
I've added a constant size and reduced this to 32 instead of 256, since nobody will have more than 32 parameters in 1 segment